### PR TITLE
prov/efa: Onboard fi_mr_dmabuf API in mem reg ops.

### PR DESCRIPTION
**Pull request by shijin-aws**
This PR contains 2 commits to make efa provider onboard fi_mr_dmabuf API in mr reg

### CODE REVIEW

**Comment by shijin-aws** [comment]
Need to rebase the PR on https://github.com/ofiwg/libfabric/pull/9499

File: prov/efa/src/efa_hmem.c [hunk 1]
```diff
@@ -148,6 +148,8 @@ static int efa_domain_hmem_info_init_cuda(struct efa_domain *efa_domain)
 	int ibv_access = IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ;
 	size_t len = ofi_get_page_size() * 2, tmp_value;
 	int ret;
+	int dmabuf_fd;
```
Comment (wenduwan) [inline prov/efa/src/efa_hmem.c:151]: nit Suggest initializing these variables.

File: prov/efa/src/efa_hmem.c [hunk 2]
```diff
@@ -174,7 +176,17 @@ static int efa_domain_hmem_info_init_cuda(struct efa_domain *efa_domain)
...
 
-	ibv_mr = ibv_reg_mr(g_device_list[0].ibv_pd, ptr, len, ibv_access);
+	ret = cuda_get_dmabuf_fd(ptr, len, &dmabuf_fd, &dmabuf_offset);
+	if (ret == FI_SUCCESS) {
+		ibv_mr = ibv_reg_dmabuf_mr(g_device_list[0].ibv_pd, dmabuf_offset,
+						len, (uint64_t)ptr, dmabuf_fd, ibv_access);
+	} else {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to retrieve dmabuf fd of CUDA device buffer, "
+			"Fall back to ibv_reg_mr\n");
```
Comment (wenduwan) [inline prov/efa/src/efa_hmem.c:186]: nit 
1. Comma -> Period
1. Suggest logging return code.

File: prov/efa/src/efa_mr.c [hunk 3]
```diff
@@ -212,6 +213,12 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 							 uint64_t flags)
 {
 	int err;
+	struct iovec mr_iov = {0};
+
+	if (flags & FI_MR_DMABUF)
+		ofi_mr_get_iov_from_dmabuf(&mr_iov, attr->dmabuf, 1);
+	else
+		memcpy(&mr_iov, attr->mr_iov, sizeof(*attr->mr_iov));
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:221]: Would it be faster to assign the fields of `mr_iov` instead of memcpy?

File: prov/efa/src/efa_mr.c [hunk 4]
```diff
@@ -934,8 +931,14 @@ static int efa_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
 	 * can be access from host. EFA provider considers all device memory
 	 * to be accessed by device only. Therefore, this function claim
 	 * support of this flag, but do not save it in efa_mr.
+	 *
+	 * FI_MR_DMABUF:
+	 * This flag indicates that the memory region to registered is
+	 * a DMA-buf backed region. When set, the region is specified through
+	 * the dmabuf field of the fi_mr_attr structure. This flag is only
+	 * usable for domains opened with FI_HMEM capability support.
 	 */
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:940]: Do we need to check API version?

File: prov/efa/src/efa_hmem.c (hunk 1 above)
Comment (shijin-aws) [reply prov/efa/src/efa_hmem.c:151]: Don't think so. `fd` and `offset` are only used when `cuda_get_dmabuf_fd` returns success, where they got assigned. We don't need to initialize them.

File: prov/efa/src/efa_mr.c [hunk 5]
```diff
@@ -202,7 +202,8 @@ static struct fi_ops efa_mr_cache_ops = {
  * efa_mr structure based on the attributes requested by the user.
  *
  * @param[in]	efa_mr	efa_mr structure to be updated
- * @param[in]	attr	fi_mr_attr from the user's registration call
+ * @param[in]	attr	a copy of fi_mr_attr updated from the user's registration call
+
```
[the remark this answers is not in the document]
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: Removed

File: prov/efa/src/efa_mr.c (hunk 4 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: We need, added

File: prov/efa/src/efa_mr.c (hunk 3 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: Updated

File: prov/efa/src/efa_hmem.c (hunk 2 above)
Comment (shijin-aws) [reply prov/efa/src/efa_hmem.c]: Fixed

File: prov/efa/src/efa_mr.c (hunk 4 above)
Comment (sunkuamzn) [reply prov/efa/src/efa_mr.c]: If we are not backporting, do we still need a version check?

Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: I think it's a different thing. The dmabuf API is only introduced in 1.20, so we don't expect user to access it when using an older Libfabric API version. They need to use FI_VERSION(1,20) in fi_getinfo to access this feature. Libfabric API version and Libfabric version are separate

Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: @j-xiong do you think we should be defensive for this api version check?

Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: Thinking about it twice, I think we need this api check, NCCL plugin will toggle dmabuf/non-dmabuf reg based on the Libfabric API version

Comment (j-xiong) [reply prov/efa/src/efa_mr.c]: Yes, I think having the version check here is recommended.

**Comment by shijin-aws** [comment]
AWS CI failure is real, need to fix the windows build

```
\libfabric\libfabric.vcxproj]
14:46:05 
14:46:05 
14:46:05        "C:\9522_1699309829\libfabric\libfabric.sln" (Build target) (1) ->
14:46:05        "C:\9522_1699309829\libfabric\libfabric.vcxproj" (default target) (5) ->
14:46:05        (Link target) -> 
14:46:05          efa_mr.obj : error LNK2001: unresolved external symbol ibv_reg_dmabuf_mr [C:\9522_1699309829\libfabric\libfabric.vcxproj]
14:46:05          C:\9522_1699309829\libfabric\x64\Release-Efa-v142\libfabric.dll : fatal error LNK1120: 1 unresolved externals [C:\9522_1699309829\libfabric\libfabric.vcxproj]
14:46:05 
14:46:05     3920 Warning(s)
14:46:05     2 Error(s)
```

File: prov/efa/src/efa_hmem.c [hunk 6]
```diff
@@ -174,7 +176,17 @@ static int efa_domain_hmem_info_init_cuda(struct efa_domain *efa_domain)
 	else
 		info->p2p_required_by_impl = true;
 
-	ibv_mr = ibv_reg_mr(g_device_list[0].ibv_pd, ptr, len, ibv_access);
+	ret = cuda_get_dmabuf_fd(ptr, len, &dmabuf_fd, &dmabuf_offset);
+	if (ret == FI_SUCCESS) {
+		ibv_mr = ibv_reg_dmabuf_mr(g_device_list[0].ibv_pd, dmabuf_offset,
+						len, (uint64_t)ptr, dmabuf_fd, ibv_access);
```
Comment (wenduwan) [inline prov/efa/src/efa_hmem.c:182]: Alignment

File: prov/efa/src/efa_mr.c [hunk 7]
```diff
@@ -251,13 +257,13 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 		efa_mr->peer.device.cuda = attr->device.cuda;
 
 		if (cuda_is_gdrcopy_enabled()) {
-			err = ofi_hmem_dev_register(FI_HMEM_CUDA, attr->mr_iov->iov_base, attr->mr_iov->iov_len,
+			err = ofi_hmem_dev_register(FI_HMEM_CUDA, mr_iov.iov_base, mr_iov.iov_len,
 							(uint64_t *)&efa_mr->peer.hmem_data);
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:261]: Alignment

File: prov/efa/src/efa_mr.c [hunk 8]
```diff
@@ -494,20 +504,38 @@ struct fi_ops efa_mr_ops = {
...
  *
  * @param efa_mr the ptr to the efa_mr object
  * @param mr_attr the ptr to the fi_mr_attr object
  * @param access the desired memory protection attributes
+ * @param flags flags in fi_mr_reg/fi_mr_regattr
  * @return struct ibv_mr* the ptr to the registered MR
  */
-static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr, int access)
+static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr,
+					 int access, const uint64_t flags)
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:518]: Alignment

File: prov/efa/src/efa_mr.c [hunk 9]
```diff
@@ -934,8 +931,23 @@ static int efa_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
...
+	 * This flag indicates that the memory region to registered is
+	 * a DMA-buf backed region. When set, the region is specified through
+	 * the dmabuf field of the fi_mr_attr structure. This flag is only
+	 * usable for domains opened with FI_HMEM capability support.
+	 * This flag is introduced since Libfabric 1.20.
 	 */
 	supported_flags = OFI_MR_NOCACHE | FI_HMEM_DEVICE_ONLY;
+
+	domain = container_of(fid, struct efa_domain,
+				util_domain.domain_fid.fid);
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:945]: Alignment

**Review by wenduwan — APPROVED** [review]
Change looks good. Formatting is off.

**Comment by j-xiong** [comment]
@shijin-aws The verbs provider has a configure variable `VERBS_HAVE_DMABUF_MR` to guard the reference of `ibv_reg_dmabuf_mr()`. This can be used by efa as well if the verbs provider is also enabled. Or we may need the check from verbs to global configure.ac. The windows config.h can then define this variable as 0 avoid any reference to the function.

The function is defined in the windows headers (prov/verbs/include/windows/infiniband/verbs.h), but the implementation is missing (prov/verbs/src/windows/verbs_nd_ibv.c). So alternatively, you could add a NOSYS implementation there since the dmabuf functionality is not supported by ND drivers.

**Comment by shijin-aws** [comment]
> @shijin-aws The verbs provider has a configure variable `VERBS_HAVE_DMABUF_MR` to guard the reference of `ibv_reg_dmabuf_mr()`. This can be used by efa as well if the verbs provider is also enabled. Or we may need to move the check from verbs to global configure.ac. The windows config.h can then define this variable as 0 avoid any reference to the function.

I will define a HAVE_EFA_DMABUF_MR macro via ac_check_decl for ibv_reg_dmabuf_mr in efa's configure.m4, and define this macro as 0 in windows/config.h

File: prov/efa/src/efa_hmem.c (hunk 6 above)
Comment (shijin-aws) [reply prov/efa/src/efa_hmem.c]: fixed

File: prov/efa/src/efa_mr.c (hunk 7 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: fixed

File: prov/efa/src/efa_mr.c (hunk 8 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: fixed

File: prov/efa/src/efa_mr.c (hunk 9 above)
Comment (shijin-aws) [reply prov/efa/src/efa_mr.c]: fixed

**Review by j-xiong — APPROVED** [review]

**Comment by j-xiong** [comment]
Intel CI error is one timeout failure in mpitestsuite with rxm/verbs. It's unrelated.
```
15:53:51  not ok 863 - ./threads/pt2pt/mt_mprobe_sendrecv_huge 2 # time=180.140264034271
```

**Pull request merged**